### ISSUE TITLE
`Selectable.asyncMap` causes latest event to not be sent to subsequent listeners

### ISSUE THREAD

**Issue by Mike278** [simolus3/moor#2443]
Hey Simon, hope you're well. 

We're doing the Drift 2.0 migration and ran into some failing tests. Here's a reduced example:
```dart
import 'package:drift/native.dart';
import 'package:drift/drift.dart';
import 'package:test/test.dart';

void main() {
  driftRuntimeOptions.dontWarnAboutMultipleDatabases = true;
  for (final suspendBetweenListeners in [true, false]) {
    for (final asyncMap in [true, false]) {
      test('suspendBetweenListeners=$suspendBetweenListeners, asyncMap=$asyncMap', () async {
        final db = TestDb();
        final select = db.customSelect('select 1');
        final stream = asyncMap
            ? select.asyncMap(Future.value).watch()
            : select.map((row) => row).watch();
        await pumpEventQueue();

        final log = [];
        stream.listen(log.add);
        if (suspendBetweenListeners) await pumpEventQueue();
        stream.listen(log.add);
        await pumpEventQueue();
        expect(log, hasLength(2));
      });
    }
  }
}

class TestDb extends GeneratedDatabase {
  TestDb() : super(NativeDatabase.memory());
  @override final List<TableInfo> allTables = const [];
  @override final int schemaVersion = 1;
}
```

The only test that fails is `suspendBetweenListeners=true, asyncMap=true`. 

Basically the real scenario is a Flutter app with two widgets listening to the same stream instance, but there's an async suspension between the 1st and 2nd `.listen`s.

Note that we're not actually using `asyncMap` for any async work - we're migrating some custom queries that take the form of:
```dart
customSelect('some gross query', variables: variables, readsFrom: readsFrom)
    .map(db.someTable.mapFromRow)
    .watch();
```
and changing `map` to `asyncMap` seemed like the simplest way forward. 

A workaround for now is to avoid asyncMap by doing:
```dart
customSelect('some gross query', variables: variables, readsFrom: readsFrom)
    .map((row) => row.data)
    .map(db.someTable.map)
    .watch();
```

**Comment by simolus3** [comment]
Thanks for the report, and I hope you're doing well too!

The issue is due to the fact that the `asyncMap` implementation from the Dart SDK uses a broadcast stream controller internally to implement the transformed stream. Broadcast stream controllers use one source for multiple listeners, so drift can't see subsequent subscriptions. This is no problem if the second listener attaches before the first query ran (because it will then receive the first result as well), but with the `pumpEventQueue()` in between the listener will be too late.

I was able to fix this by rolling a custom `asyncMap` implementation that does the conversion for each subscription (which is also the way `Stream.map` is implemented). I'll point to https://github.com/dart-lang/tools/issues/1778 and argue that `asyncMap` should have been implemented similarly, but that's unlikely to change now.

-------

A somewhat unfortunate consequence of this is other transformations in the Dart SDK and `rxdart` likely have the same problem (and IIRC we've seen a few of those already). But at least that's somewhat easier to understand - after all, the behavior is not unexpected when wrapping streams in a broadcast controller. IMHO the classification of streams into "single-subscription only" and "same behavior for all listeners" given by `Stream.isBroadcast` is pretty unfortunate (a package like `rxdart` has to roll similar special cases for their `BehaviorSubject`). The important part is that drift APIs can transform drift streams correctly.

**Comment by Mike278** [comment]
Oh jeez yeah I remember that `switchMap` issue. Definitely an unfortunate situation, and as natebosch pointed out you can reproduce the bug solely using Dart SDK transformations - `map` uses a `_ForwardingStream` subclass and `asyncMap` uses a broadcast stream controller. "it's how broadcast stream controllers work" also isn't really a satisfying answer because it seems like that should be an implementation detail (hence the "Not all current stream modifiers do that, it was something we wanted to fix for Dart 2.0, but didn't have time for"). 

Anyway, thanks for getting back to me so quick! That commit mostly fixes it, but there are 2 other `asyncMap` usages in [select.dart](https://github.com/simolus3/drift/blob/3b68386992c0dbdb1256391ac3668c9981831328/drift/lib/src/runtime/query_builder/statements/select/select.dart#L76) and [select_with_join.dart](https://github.com/simolus3/drift/blob/3b68386992c0dbdb1256391ac3668c9981831328/drift/lib/src/runtime/query_builder/statements/select/select_with_join.dart#L236) that are still causing some problems.

**Comment by simolus3** [comment]
Good catch, thanks! Those are fixed as well in 01771758789070c9deb8401fe72b925a7087dba6 (which also fixes a nasty regression - the custom `asyncMap` streams didn't forward cancellation events).
